api/organization.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+)
+
+//OrganizationMiddleware parses the organization id from the request, queries it from the database and saves it to the current context
+func OrganizationMiddleware(c *gin.Context) {
+	log := logger.WithFields(logrus.Fields{"tag": "OrganizationMiddleware"})
+	orgidParam := c.Param("orgid")
+	orgid, err := strconv.ParseUint(orgidParam, 10, 32)
+	if err != nil {
+		message := fmt.Sprintf("error parsing organization id: %q", orgidParam)
+		log.Info(message)
+		c.AbortWithStatusJSON(http.StatusBadRequest, components.ErrorResponse{
+			Code:    http.StatusBadRequest,
+			Message: message,
+			Error:   message,
+		})
+	}
+
+	user := auth.GetCurrentUser(c.Request)
+	var organization = auth.Organization{ID: uint(orgid)}
+	var organizations []auth.Organization
+
+	db := model.GetDB()
+	err = db.Model(user).Where(&organization).Related(&organizations, "Organizations").Error
+	if err != nil {
+		message := "error fetching organizations"
+		log.Info(message + ": " + err.Error())
+		c.AbortWithStatusJSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
+		})
+		return
+	}
+
+	if len(organizations) != 1 {
+		message := fmt.Sprintf("organization not found: %q", orgidParam)
+		log.Info(message)
+		c.AbortWithStatusJSON(http.StatusNotFound, components.ErrorResponse{
+			Code:    http.StatusNotFound,
+			Message: message,
+			Error:   message,
+		})
+		return
+	}
+	newContext := context.WithValue(c.Request.Context(), auth.CurrentOrganization, &organizations[0])
+	c.Request = c.Request.WithContext(newContext)
+	c.Next()
+}
+
+//GetOrganizations returns all organizations the user belongs to or a specific one from those by id
+func GetOrganizations(c *gin.Context) {
+	log := logger.WithFields(logrus.Fields{"tag": "GetOrganizations"})
+	log.Info("Fetching organizations")
+
+	user := auth.GetCurrentUser(c.Request)
+
+	idParam := c.Param("id")
+	id, err := strconv.ParseUint(idParam, 10, 32)
+	if idParam != "" && err != nil {
+		message := fmt.Sprintf("error parsing organization id: %s", err)
+		log.Info(message)
+		c.JSON(http.StatusBadRequest, components.ErrorResponse{
+			Code:    http.StatusBadRequest,
+			Message: message,
+			Error:   message,
+		})
+		return
+	}
+
+	var organization = auth.Organization{ID: uint(id)}
+	var organizations []auth.Organization
+	db := model.GetDB()
+	err = db.Model(user).Where(&organization).Related(&organizations, "Organizations").Error
+	if err != nil {
+		message := "error fetching organizations"
+		log.Info(message + ": " + err.Error())
+		c.JSON(http.StatusInternalServerError, components.ErrorResponse{
+			Code:    http.StatusInternalServerError,
+			Message: message,
+			Error:   message,
+		})
+	} else if id == 0 {
+		c.JSON(http.StatusOK, organizations)
+	} else if len(organizations) != 0 {
+		c.JSON(http.StatusOK, organizations[0])
```

Comment:
If an `id` is passed in that there should be exactly one match. If more than one organisation is found than that is an error case.